src/backend/distributed/commands/multi_copy.c

```diff
@@ -2923,3 +3041,308 @@ CopyGetAttnums(TupleDesc tupDesc, Relation rel, List *attnamelist)
+	{
+		InitializeCopyShardState(shardState, connectionStateHash, shardId);
+	}
+
+	return shardState;
+}
+
+
+/*
+ * InitializeCopyShardState initializes the given shardState. It finds all
+ * placements for the given shardId, assignes connections to them, and
+ * adds them to shardState->placementStateList.
+ */
+static void
+InitializeCopyShardState(CopyShardState *shardState,
+						 HTAB *connectionStateHash, uint64 shardId)
+{
+	List *finalizedPlacementList = NIL;
+	ListCell *placementCell = NULL;
+
+	MemoryContext localContext =
+		AllocSetContextCreateExtended(CurrentMemoryContext,
+									  "InitializeCopyShardState",
+									  ALLOCSET_DEFAULT_MINSIZE,
+									  ALLOCSET_DEFAULT_INITSIZE,
+									  ALLOCSET_DEFAULT_MAXSIZE);
+
+	/* release finalized placement list at the end of this function */
+	MemoryContext oldContext = MemoryContextSwitchTo(localContext);
+
+	finalizedPlacementList = MasterShardPlacementList(shardId);
+
+	MemoryContextSwitchTo(oldContext);
+
+	shardState->shardId = shardId;
+	shardState->placementStateList = NIL;
+
+	foreach(placementCell, finalizedPlacementList)
+	{
+		ShardPlacement *placement = (ShardPlacement *) lfirst(placementCell);
+		MultiConnection *connection = CopyGetPlacementConnection(placement);
+		CopyConnectionState *connectionState = GetConnectionState(connectionStateHash,
+																  connection);
+		CopyPlacementState *placementState = palloc0(sizeof(CopyPlacementState));
+		placementState->shardState = shardState;
+		placementState->data = makeStringInfo();
+		placementState->connectionState = connectionState;
+		dlist_push_head(&connectionState->bufferedPlacementList,
+						&placementState->bufferedPlacementNode);
+		shardState->placementStateList = lappend(shardState->placementStateList,
+												 placementState);
+	}
+
+	MemoryContextReset(localContext);
+}
+
+
+/*
+ * CopyGetPlacementConnection assigns a connection to the given placement. If
+ * a connection has already been assigned the placement in the current transaction
+ * then it reuses the connection. Otherwise, it requests a connection for placement.
+ */
+static MultiConnection *
+CopyGetPlacementConnection(ShardPlacement *placement)
+{
+	MultiConnection *connection = NULL;
+	uint32 connectionFlags = FOR_DML;
+	char *nodeUser = CurrentUserName();
+	List *placementAccessList = NIL;
+	ShardPlacementAccess *placementAccess = NULL;
+
+	/*
+	 * Determine whether the task has to be assigned to a particular connection
+	 * due to a preceding access to the placement in the same transaction.
+	 */
+	placementAccess = CreatePlacementAccess(placement, PLACEMENT_ACCESS_DML);
+	connection = GetPlacementListConnectionIfCached(connectionFlags,
+													list_make1(placementAccess),
+													NULL);
+	if (connection != NULL)
+	{
+		return connection;
+	}
+
+	/*
+	 * Make sure we use a separate connection per placement for hash-distributed
+	 * tables in order to allow multi-shard modifications in the same transaction.
+	 */
```

Comment:
We should probably update the comment to say that we now do this to maximize performance for large COPY batches.

Reply:
done